### PR TITLE
[CALCITE-5976] Use explicit casting if inserted element type in ArrayPrepend/ArrayAppend/ArrayInsert does not equal derived component type

### PR DESCRIPTION
**Pull request by caicancai**
https://issues.apache.org/jira/browse/CALCITE-5976

### CODE REVIEW

**Comment by caicancai** [comment]
@chucheng92 Hello, I'm not done yet, I want to make sure the way I'm fixing it is acceptable

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -5204,7 +5204,11 @@ public static List compact(List list) {
   public static List arrayAppend(List list, Object element) {
     final List result = new ArrayList(list.size() + 1);
     result.addAll(list);
-    result.add(element);
+    if (element instanceof Byte) {
+      result.add(((Byte) element).intValue());
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:5208]: This doesn't seem right. Is the array always containing integers?
What happens for an array of BIGINT?
What happens for Short?

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:5208]: Thanks I will try other methods

Comment (chucheng92) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:5208]: +1 can't do this. we must make sure when you call `arrayAppend` the input `List` already is a isomorphic result list (a result list has same operand's type). so it means we need an explicit casting before you call SqlFunctions.

**Pull request closed without merging**